### PR TITLE
Extract and render output values from Terraform

### LINKED ISSUE #165: Inject `terraform output` into `terraform-docs` as an additional 'Value' field
### Prerequisites

Put an `x` into the box that applies:

- [ ] This issue describes a bug.
- [x] This issue describes a feature request.

### Description
I'd like to have `terraform-docs` add an optional field for Outputs which would specify the current value of an output variable as it's found in state via `terraform output`.

My use-case:
The outputs of my Terraform project get imported into the Terraform projects of my users as variables. Ideally, my documentation would inform users what value to expect when they use them.

Let me know if any of that could use further clarification or if you have any feedback or suggestions.

Cheers!

### PR DESCRIPTION
**Pull request by gshel**
### Prerequisites

Put an `x` into the box(es) that apply:

- [ ] This pull request fixes a bug.
- [X] This pull request adds a feature.
- [ ] This pull request enhances existing functionality.
- [ ] This pull request introduces breaking change.

For more information, see the [Contributing Guide](https://github.com/segmentio/terraform-docs/tree/master/CONTRIBUTING.md).

### Description

This PR allows users to pass the global flag '--output-values PATH' to inject a project's outputs' values into the chosen format.

1. `cd intoYourTF/project`
2. `terraform outputs --json > output_values.json`
3. `terraform-docs markdown table --output-values output_values.json`

### Issues Resolved
#165 

### Checklist

Put an `x` into all boxes that apply:

- [x] I have read the [Contributing Guidelines](https://github.com/segmentio/terraform-docs/tree/master/CONTRIBUTING.md).

#### Tests

- [x] I have added tests to cover my changes.
- [x] All tests pass when I run `make test`.

#### Documentation

- [x] My change requires a change to the documentation.
- [TBD] I have updated the documentation accordingly.

#### Code Style

- [x] My code follows the code style of this project.

### CODE REVIEW

**Comment by gshel** [comment]
Some things I can think off of the top of my head that I would like to hear your input on:

- If an output value is `""` should that display as `n/a`?

- If an output value is `""` should json/yaml omit the empty field?

- I looked over the documentation, noticing that there are not currently sections detailing flags. Since you marked my issue with the `documentation` label, I wanted to confirm that you'd like it added into the readme.

- I have some clunky if/else statements that I'm wondering if there are alternative, elegant solutions for (e.g. `internal/pkg/tfconf/module.go:142`)

File: internal/pkg/print/pretty/pretty.go [hunk 1]
```diff
@@ -52,8 +52,7 @@ const (
 			{{- printf "\n" -}}
 			{{- range . }}
 				{{ printf "output.%s" .Name | colorize "\033[36m" }}
-				{{ tostring .Description | trimSuffix "\n" | default "n/a" | colorize "\033[90m" }}
-				{{ if $.Settings.OutputValues }}{{ .Value | trimSuffix "\n" | default "n/a" | colorize "\033[90m" }}{{ end}}
+				{{ tostring .Description | trimSuffix "\n" | default "n/a" | colorize "\033[90m" }}{{ if $.Settings.OutputValues }}{{- printf "\n" -}}{{ .Value | trimSuffix "\n" | default "n/a" | colorize "\033[90m" }}{{ end }}
```
Comment (gshel) [inline internal/pkg/print/pretty/pretty.go]: So ugly, but needed to keep the `pretty` output the same.

File: cmd/root.go [hunk 2]
```diff
@@ -51,6 +54,8 @@ func init() {
 	rootCmd.PersistentFlags().BoolVar(new(bool), "no-sort", false, "do no sort items")
 	rootCmd.PersistentFlags().BoolVar(&settings.SortByRequired, "sort-by-required", false, "sort items by name and print required ones first")
 
+	rootCmd.PersistentFlags().StringVar(&outputValuesPath, "output-values", "", "inject output values from `terraform output --json` into outputs")
```
Comment (khos2ow) [inline cmd/root.go]: Ok now I see your point about this new flag, I then suggest to split that into two flags:

- `--output-values bool` : to indicate if output values need to get extract or not
- `--output-values-from string` : path of the json file to read the output from (defaults to `terraform-output.json`

Then we would be able to internally use terraform binary itself to extract json of output and parse it rather than forcing one extra step to the user. Although sometime a user might choose to do that extra step manually for whatever reason which will be fine too.

File: cmd/root.go [hunk 3]
```diff
@@ -74,7 +79,7 @@ func Execute() error {
 }
 
 func doPrint(paths []string, fn func(*tfconf.Module) (string, error)) {
-	module, err := tfconf.CreateModule(paths[0])
+	module, err := tfconf.CreateModule(paths[0], outputValuesPath)
```
Comment (khos2ow) [inline cmd/root.go]: Since we're starting to add more variables to `CreateModule` function I'd suggest we create a struct of its own, and have that as the only argument for `CreateModule(option Option)`. something like, `module.Option` in `tfconf/options.go` which has the following:

- `Path string`
- `OutputValues bool`
- `OutputValuesFile string`

File: internal/pkg/print/markdown/table/table.go [hunk 4]
```diff
@@ -64,10 +64,10 @@ const (
 		{{ if not .Module.Outputs }}
 			No output.
 		{{ else }}
-			| Name | Description |
-			|------|-------------|
+			| Name | Description | {{if $.Settings.OutputValues}}Value |{{end}}
```
Comment (khos2ow) [inline internal/pkg/print/markdown/table/table.go]: Please move the leading space from before {{if &.Settings....` to inside the if.

And also put surrounding spaces inside `{{ ... }}`, e.g. `{{ end }}` instead of `{{end}}`.

File: internal/pkg/print/markdown/table/table.go [hunk 5]
```diff
@@ -64,10 +64,10 @@ const (
...
+			|------|-------------|{{if $.Settings.OutputValues}}-------|{{end}}
```
Comment (khos2ow) [inline internal/pkg/print/markdown/table/table.go]: Same surrounding spaces.

File: internal/pkg/print/markdown/table/table.go [hunk 6]
```diff
@@ -64,10 +64,10 @@ const (
...
 			{{- range .Module.Outputs }}
-				| {{ name .Name }} | {{ tostring .Description | sanitizeTbl }} |
+				| {{ name .Name }} | {{ tostring .Description | sanitizeTbl }} | {{if $.Settings.OutputValues}}{{ .Value | sanitizeTbl }} |{{end}}
```
Comment (khos2ow) [inline internal/pkg/print/markdown/table/table.go]: Same as above again.

File: internal/pkg/testutil/testing.go [hunk 7]
```diff
@@ -16,7 +16,7 @@ func GetExpected(goldenFile string) (*tfconf.Module, string, error) {
 		return nil, "", err
 	}
 
-	module, err := tfconf.CreateModule(path)
+	module, err := tfconf.CreateModule(path, "")
```
Comment (khos2ow) [inline internal/pkg/testutil/testing.go]: See my note above for `module.Option`.

File: internal/pkg/print/pretty/pretty.go [hunk 8]
```diff
@@ -52,7 +52,7 @@ const (
 			{{- printf "\n" -}}
 			{{- range . }}
 				{{ printf "output.%s" .Name | colorize "\033[36m" }}
-				{{ tostring .Description | trimSuffix "\n" | default "n/a" | colorize "\033[90m" }}
+				{{ tostring .Description | trimSuffix "\n" | default "n/a" | colorize "\033[90m" }}{{ if $.Settings.OutputValues }}{{- printf "\n" -}}{{ .Value | trimSuffix "\n" | default "n/a" | colorize "\033[90m" }}{{ end }}
```
Comment (khos2ow) [inline internal/pkg/print/pretty/pretty.go]: To keep it consistent with input values, I think this would work better. this will show the output value in `(...)` in front of the output name and without any color, the same as inputs.

```suggestion
				{{- if $.Settings.OutputValues -}}
					{{- printf " " -}}
					({{ .Value | trimSuffix "\n" | default "n/a" }})
				{{- end }}
				{{ tostring .Description | trimSuffix "\n" | default "n/a" | colorize "\033[90m" }}
```

File: internal/pkg/tfconf/output.go [hunk 9]
```diff
@@ -4,6 +4,7 @@ package tfconf
 type Output struct {
 	Name        string   `json:"name" yaml:"name"`
 	Description String   `json:"description" yaml:"description"`
+	Value       string   `json:"value,omitempty" yaml:"value,omitempty"`
```
Comment (khos2ow) [inline internal/pkg/tfconf/output.go]: This will have some limitation as it will only render the value as `string` and also omit them if they are empty, which means the schema of json will be different based on the selected flags. You don't need to worry about this, we will have to circle back to it when the above comments are done.

Also I still need to understand how `terraform output --json` exactly works, I've briefly tested something and got different results based on the version of terraform binary I used.

**Comment by khos2ow** [comment]
> * If an output value is `""` should that display as `n/a`?
> * If an output value is `""` should json/yaml omit the empty field?

We definitely should not drop the empty field, I don't have an answer for it at the moment, check out https://github.com/segmentio/terraform-docs/pull/191#discussion_r376729640

> * I looked over the documentation, noticing that there are not currently sections detailing flags. Since you marked my issue with the `documentation` label, I wanted to confirm that you'd like it added into the readme.

There are some documentation available [here](https://github.com/segmentio/terraform-docs/tree/master/docs/formats). I'm currently working on an enhancement to auto-generate that from whatever exists in [`cmd/`](https://github.com/segmentio/terraform-docs/tree/master/cmd). I put the `documentation` label to not forget about it.

> * I have some clunky if/else statements that I'm wondering if there are alternative, elegant solutions for (e.g. `internal/pkg/tfconf/module.go:142`)

That's true :) we need to address that after/with this https://github.com/segmentio/terraform-docs/pull/191#discussion_r376729640

File: internal/pkg/print/pretty/pretty.go (hunk 1 above)
Comment (khos2ow) [reply internal/pkg/print/pretty/pretty.go]: Please take a look at https://github.com/segmentio/terraform-docs/pull/191#discussion_r376729470

File: internal/pkg/tfconf/output.go (hunk 9 above)
Comment (gshel) [reply internal/pkg/tfconf/output.go]: When you run the latest terraform, are you getting errors like I mentioned here?  #183 
 https://github.com/segmentio/terraform-docs/issues/183

File: internal/pkg/print/pretty/pretty.go (hunk 8 above)
Comment (gshel) [reply internal/pkg/print/pretty/pretty.go]: - [X] Make line 55 prettier

File: internal/pkg/print/markdown/table/table.go (hunk 6 above)
Comment (gshel) [reply internal/pkg/print/markdown/table/table.go]: - [X] Fix spacing

File: internal/pkg/print/markdown/table/table.go (hunk 5 above)
Comment (gshel) [reply internal/pkg/print/markdown/table/table.go]: - [X] Fix Spacing

File: internal/pkg/print/markdown/table/table.go (hunk 4 above)
Comment (gshel) [reply internal/pkg/print/markdown/table/table.go]: - [X] Fix Spacing

File: cmd/root.go (hunk 3 above)
Comment (khos2ow) [reply cmd/root.go]: I've just pushed #193, you can take a look and let me know. We can proceed with that one first and then you rebase, or we can proceed with this PR (disregarding my comment above) and then I'll rebase the other one. Feel free to let me know which one you prefer.

File: internal/pkg/testutil/testing.go (hunk 7 above)
Comment (khos2ow) [reply internal/pkg/testutil/testing.go]: This also can be disregarded considering #193

File: internal/pkg/tfconf/output.go (hunk 9 above)
Comment (khos2ow) [reply internal/pkg/tfconf/output.go]: No it wasn't like #183 I was getting different `type` and `value`. Some of them were plain `string` and `"some value"`, some other were `["list", "string"]` and `["val1", "val2"]` and more. I haven't had enough time to try to understand a pattern.

File: cmd/root.go (hunk 2 above)
Comment (gshel) [reply cmd/root.go]: I gave this some thought. What do you think of this alternative?

User passes `--output-values PATH` to specify a file from which to take output values.
OR
User passes `--output-values` to use the default PATH of `terraform-output.json` which is output by `terraform output --json > terraform-output.json`

The way I currently have set up has the default PATH set to an empty string, but I could easily change that to `terraform-output.json` and say that if the variable `outputValuesPath` is `""`, execute `terraform output --json > terraform-output.json`.

Two flags to do essentially the same thing sounds a bit :stuck_out_tongue_closed_eyes: to me, but it's entirely subjective.

File: cmd/root.go (hunk 3 above)
Comment (gshel) [reply cmd/root.go]: Looks like yours is a lot more complete than what I've done for this portion. I can rebase after you merge :smile_cat:

Comment (khos2ow) [reply cmd/root.go]: Cool, so I'm gonna merge that and please rebase here and add required properties to the new `Options`.

File: internal/pkg/tmpl/template.go [hunk 10]
```diff
@@ -160,6 +160,24 @@ func builtinFuncs(settings *print.Settings) map[string]interface{} {
 		"sanitizeTbl": func(s string) string {
 			return markdown.SanitizeItemForTable(s, settings)
 		},
```
Comment (gshel) [inline internal/pkg/tmpl/template.go]: @khos2ow How do you feel about the below changes? This is how I was able to keep the type in the generated print out. If there are any cases I have forgotten, let me know!

File: internal/pkg/tfconf/module.go [hunk 11]
```diff
@@ -124,24 +126,35 @@ func CreateModule(options *Options) (*Module, error) {
 		}
 	}
```
Comment (gshel) [inline internal/pkg/tfconf/module.go]: I will change the env variable name if this ends up being an accepted solution - suggestions welcome!

File: internal/pkg/tfconf/output.go [hunk 12]
```diff
@@ -1,11 +1,29 @@
 package tfconf
 
+import (
+	"encoding/json"
+)
+
 // Output represents a Terraform output.
 type Output struct {
```
Comment (gshel) [inline internal/pkg/tfconf/output.go]: By changing the Value attribute's type to `interface{}`, we are able to include typing in the generated printouts.

File: cmd/root.go (hunk 2 above)
Comment (khos2ow) [reply cmd/root.go]: To be fair the two flags I mentioned don't do the same thing, `--output-values bool` is the indicator that we want to have outputs processed, and `--output-values-from string` is to specify from which JSON file we should read them, otherwise internally using `terraform` binary to attempt to fetch it within the execution of terraform-docs command.

File: cmd/root.go [hunk 13]
```diff
@@ -88,7 +90,8 @@ func FormatterCmds() []*cobra.Command {
 
 func doPrint(path string, fn func(*tfconf.Module) (string, error)) {
 	options := &tfconf.Options{
-		Path: path,
+		Path:         path,
+		OutputValues: settings.OutputValues,
```
Comment (khos2ow) [inline cmd/root.go]: You actually defined `OutputValuesPath string` in `tfconf.Options` but don't pass it from here. I'd suggest to create a corresponding flag for it and pass it down.

File: internal/pkg/print/markdown/table/table.go [hunk 14]
```diff
@@ -64,10 +64,10 @@ const (
 		{{ if not .Module.Outputs }}
 			No output.
 		{{ else }}
-			| Name | Description |
-			|------|-------------|
+			| Name | Description |{{ if $.Settings.OutputValues }} Value |{{ end }}
+			|------|-------------|{{ if $.Settings.OutputValues }}-------|{{ end }}
 			{{- range .Module.Outputs }}
-				| {{ name .Name }} | {{ tostring .Description | sanitizeTbl }} |
+				| {{ name .Name }} | {{ tostring .Description | sanitizeTbl }} |{{ if $.Settings.OutputValues }} {{ .Value | sanitizeInterface | sanitizeTbl }} | {{ end }}
```
Comment (khos2ow) [inline internal/pkg/print/markdown/table/table.go]: There's an extra space right before the last `| {{ end }}` which should be removed (that's a markdown lint warning)

File: internal/pkg/print/settings.go [hunk 15]
```diff
@@ -14,6 +14,10 @@ type Settings struct {
 	// scope: Markdown
 	MarkdownIndent int
 
+	// OutputValues inject output values from `terraform output --json` into outputs (default: false)
+	// scopt: Global
+	OutputValues bool
```
Comment (khos2ow) [inline internal/pkg/print/settings.go]: I'm thinking we don't need this flag here since it's already defined in `tfconf.Options` and this is in fact an _option_ to load the module with not a _setting_ to render and format result.

File: internal/pkg/tfconf/module.go [hunk 16]
```diff
@@ -122,20 +126,57 @@ func CreateModule(options *Options) (*Module, error) {
...
 
+	options.OutputValuesPath = os.Getenv("FOO")
+	// if OutputValuesPath is empty, set the default and generate the file
+	if options.OutputValues && options.OutputValuesPath == "" {
+		err := os.Setenv("FOO", "terraform-outputs.json")
```
Comment (khos2ow) [inline internal/pkg/tfconf/module.go]: I'm completely against the use of environment variable. Actually to rephrase if environment variable should be used they should be immediately parsed right from the get go (a.k.a inside `cmd/root.go`) and the parsed values should be passed down.

But in this particular case I cannot think of a reason where we should need to use this. Having a flag is extremely visible and self-explanatory but environment variables are not and error prone. Also autocomplete works for the flags and not the environment variables.

File: internal/pkg/tmpl/template.go (hunk 10 above)
Comment (khos2ow) [reply internal/pkg/tmpl/template.go]: I'm gonna take a closer look, but by the look of it it seems great. Although I'm also working on another PR to streamline the use of "default values" and not actually depend on `interface{}` as a literal. I'm gonna have to combine these change of yours and mine together to see if we would need both or either. I'm gonna circle back on this soon.

File: internal/pkg/tfconf/module.go [hunk 17]
```diff
@@ -122,20 +126,57 @@ func CreateModule(options *Options) (*Module, error) {
...
+		if err != nil {
+			log.Fatal(err)
+		}
+		_, err = exec.Command("cd", options.Path, "&&", "terraform", "output", "--json", ">>", options.OutputValuesPath).Output()
```
Comment (khos2ow) [inline internal/pkg/tfconf/module.go]: This looks the way I had thought, I'm gonna take a close look shortly.

File: internal/pkg/tfconf/module.go (hunk 11 above)
Comment (khos2ow) [reply internal/pkg/tfconf/module.go]: Please look at https://github.com/segmentio/terraform-docs/pull/191#discussion_r379701094

File: internal/pkg/print/markdown/table/table.go (hunk 14 above)
Comment (gshel) [reply internal/pkg/print/markdown/table/table.go]: Okay, fixing that now. Btw sorry if my PR has become a bit of a clusterf. I'm not super accustomed to squashing/resolving merge conflicts :stuck_out_tongue: Will make changes for this shortly!

Comment (khos2ow) [reply internal/pkg/print/markdown/table/table.go]: No, it's completely find, not squashing commits and PR of this size is actually a better thing, makes with code review. And I'll squash merge the PR anyways, so no worries!

File: internal/pkg/tfconf/module.go [hunk 18]
```diff
@@ -122,20 +126,57 @@ func CreateModule(options *Options) (*Module, error) {
...
+	var terraformOutputs map[string]*TerraformOutput
+	if options.OutputValues {
+		byteValue, err := ioutil.ReadFile(options.OutputValuesPath)
+		if err != nil {
+			return nil, fmt.Errorf("caught error while reading the terraform outputs file at %s: %v", options.OutputValuesPath, err)
+		}
+		if err := json.Unmarshal(byteValue, &terraformOutputs); err != nil {
+			return nil, err
+		}
+	}
```
Comment (khos2ow) [inline internal/pkg/tfconf/module.go]: Please extract these into a function, like `loadOutputValues`. The flow of code will be cleaner and it will be easier to test individual function and helps for mocking (if need be)

File: internal/pkg/tfconf/module.go [hunk 19]
```diff
@@ -122,20 +126,57 @@ func CreateModule(options *Options) (*Module, error) {
...
+			outputs = append(outputs, &Output{
+				Name:        output.Name,
+				Description: String(outputDescription),
+				Position: Position{
+					Filename: output.Pos.Filename,
+					Line:     output.Pos.Line,
+				},
+			})
+		}
 	}
```
Comment (khos2ow) [inline internal/pkg/tfconf/module.go]: This looks cleaner and more readable to me:

```
	for _, o := range mod.Outputs {
		outputDescription := o.Description
		if outputDescription == "" {
			outputDescription = readComment(o.Pos.Filename, o.Pos.Line-1)
		}
		output := &Output{
			Name:        o.Name,
			Description: String(outputDescription),
			Position: Position{
				Filename: o.Pos.Filename,
				Line:     o.Pos.Line,
			},
		}
		if options.OutputValues {
			output.Value = terraformOutputs[output.Name].Value
		}
		outputs = append(outputs, output)
	}
```

File: internal/pkg/tfconf/module.go (hunk 11 above)
Comment (gshel) [reply internal/pkg/tfconf/module.go]: Removed the env variable, let me know what you think~

File: internal/pkg/tfconf/module.go (hunk 16 above)
Comment (gshel) [reply internal/pkg/tfconf/module.go]: Refactored with your original suggestions!

File: internal/pkg/testutil/testing.go [hunk 20]
```diff
@@ -1,34 +1,38 @@
...
 	"os"
 	"path/filepath"
 	"runtime"
-
-	"github.com/segmentio/terraform-docs/internal/pkg/tfconf"
 )
 
-// GetExpected returns 'example' Module and expected Golden file content
-func GetExpected(goldenFile string) (*tfconf.Module, string, error) {
-	path, err := getExampleFolder()
```
Comment (gshel) [inline internal/pkg/testutil/testing.go]: I wasn't expecting the `GetExpected` function to also generate the module, so I broke it into two separate functions:

1. GetExpected: Read the Golden File and assign to expected
2. GetModule: Accept options and generate the module used for the `actual` var

File: internal/pkg/tfconf/module.go (hunk 19 above)
Comment (gshel) [reply internal/pkg/tfconf/module.go]: Thanks!

File: internal/pkg/tfconf/module.go (hunk 18 above)
Comment (gshel) [reply internal/pkg/tfconf/module.go]: All done~

File: internal/pkg/print/settings.go (hunk 15 above)
Comment (gshel) [reply internal/pkg/print/settings.go]: Removed from settings :heavy_check_mark:

File: cmd/root.go (hunk 13 above)
Comment (gshel) [reply cmd/root.go]: Fixed this~

File: cmd/root.go (hunk 2 above)
Comment (gshel) [reply cmd/root.go]: Refactored with your suggestions~

**Comment by gshel** [comment]
I'm not sure how we can test `--outputValues` without including the path to my pre-generated `output_values.json`. I ran into issues running `terraform output --json` inside the examples folder because I'm using Terraform v0.12.19. Ideas?

File: internal/pkg/testutil/testing.go (hunk 20 above)
Comment (khos2ow) [reply internal/pkg/testutil/testing.go]: Yes, I've noticed this and have already had it converted to exactly like what you did in #195! Nice work.

File: internal/pkg/testutil/testing.go [hunk 21]
```diff
@@ -9,26 +9,31 @@ import (
 	"github.com/segmentio/terraform-docs/internal/pkg/tfconf"
 )
 
-// GetExpected returns 'example' Module and expected Golden file content
-func GetExpected(goldenFile string) (*tfconf.Module, string, error) {
-	path, err := getExampleFolder()
+// GetModule returns what is generated by the test params
+func GetModule(options *tfconf.Options) (*tfconf.Module, error) {
+	path, err := GetExampleFolder()
```
Comment (khos2ow) [inline internal/pkg/testutil/testing.go]: Although I didn't get why you made `GetExampleFolder` exported since it's only being used in `internal/pkg/testutil/testing.go`. I'd suggest to revert it back to `getExampleFolder` (effectively make it private)

File: internal/pkg/testutil/testing.go [hunk 22]
```diff
@@ -9,26 +9,31 @@ import (
...
 	if err != nil {
-		return nil, "", err
+		return nil, err
 	}
+	return module, err
+}
 
+// GetExpected returns 'example' Module and expected Golden file content
+func GetExpected(goldenFile string) (string, error) {
 	expected, err := readGoldenFile(goldenFile)
```
Comment (khos2ow) [inline internal/pkg/testutil/testing.go]: Also having a separate `readGoldenFile` seems to be irrelevant and its content can be moved here. something like this:

```go
// GetExpected returns 'example' Module and expected Golden file content
func GetExpected(format, name string) (string, error) {
	path := filepath.Join(testDataPath(), format, name+".golden")
	bytes, err := ioutil.ReadFile(path)
	if err != nil {
		return "", err
	}
	return string(bytes), nil
}
```
and delete `readGoldenFile` function altogether.

File: internal/pkg/tfconf/output.go (hunk 12 above)
Comment (khos2ow) [reply internal/pkg/tfconf/output.go]: I just opened PR #196 which addresses this (well `Input`, but this is similar). I think it's fine we keep this change of yours (i.e. using `interface{}`) as is here and I'll rebase mine later when we get this PR merged. :+1:

File: internal/pkg/tmpl/template.go (hunk 10 above)
Comment (khos2ow) [reply internal/pkg/tmpl/template.go]: As my above comment, I think this is good for now and we can take required actions if needed in #196 after rebase.

File: internal/pkg/tfconf/module.go [hunk 23]
```diff
@@ -22,8 +25,7 @@ type Module struct {
 	Providers      []*Provider `json:"providers" yaml:"providers"`
 	RequiredInputs []*Input    `json:"-" yaml:"-"`
 	OptionalInputs []*Input    `json:"-" yaml:"-"`
-
-	options *Options
+	Options        *Options    `json:"-" yaml:"-"`
```
Comment (khos2ow) [inline internal/pkg/tfconf/module.go]: Ok I see why you had to make `Options` here exported, we shouldn't do that, because `Module` will eventually published publicly and consumer shouldn't have access to it (look #195 for detail if interested). It was my bad, I originally suggested to completely remove `OutputValues bool` from `print.Settings` which we should move it back.

And to be clear we need `OutputValues bool` in both `print.Settings` and `tfconf.Options`, because I'm trying to put a boundaries between the two in #195 . I mean `tfconf` package must not have dependency to `print` package.

File: internal/pkg/print/pretty/pretty.go [hunk 24]
```diff
@@ -52,7 +52,11 @@ const (
 			{{- printf "\n" -}}
 			{{- range . }}
 				{{ printf "output.%s" .Name | colorize "\033[36m" }}
-				{{ tostring .Description | trimSuffix "\n" | default "n/a" | colorize "\033[90m" }}
+				{{- if $.Module.Options.OutputValues -}}
```
Comment (khos2ow) [inline internal/pkg/print/pretty/pretty.go]: Please take a look at my comment below regarding `Settings` and `Options` and not exporting `Module.Options` and use `$.Settings.OutputValues` here.

File: internal/pkg/tfconf/module.go [hunk 25]
```diff
@@ -187,3 +198,24 @@ func loadProviders(requiredProviders map[string]*tfconfig.ProviderRequirement, r
...
 	return providers
 }
+
+func loadOutputValues(options *Options) (map[string]*TerraformOutput, error) {
+	if options.OutputValuesPath == "terraform-outputs.json" {
+		_, err := exec.Command("cd", options.Path, "&&", "terraform", "output", "--json", ">>", options.OutputValuesPath).Output()
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
```
Comment (khos2ow) [inline internal/pkg/tfconf/module.go]: This is the only remaining items, from where I can see. I'm gonna test this locally today. But my initial thought at the moment is that we actually should not set a default value for `--output-values-from`, but let me gather my thoughts on it first. Will sync up with you shortly.

File: internal/pkg/print/settings.go (hunk 15 above)
Comment (khos2ow) [reply internal/pkg/print/settings.go]: Please take a look at https://github.com/segmentio/terraform-docs/pull/191#discussion_r380339458 , this needs to get reverted.

File: internal/pkg/testutil/testing.go (hunk 21 above)
Comment (gshel) [reply internal/pkg/testutil/testing.go]: Oops leftover from something else I had tried. Thanks for catching that.

File: internal/pkg/testutil/testing.go (hunk 22 above)
Comment (gshel) [reply internal/pkg/testutil/testing.go]: Was wondering what the `format` var is meant to be? Previously, `GetExpected` was taking the golden file name less the extension as the arg.

Comment (khos2ow) [reply internal/pkg/testutil/testing.go]: That was copy paste issue :see_no_evil: which we will need in the PR #195 which refactors stuff. For this PR only `name` suffices.

Comment (gshel) [reply internal/pkg/testutil/testing.go]: :smile:  sounds good! I figured, but wanted to clarify. Will be pushing shortly!

File: internal/pkg/tfconf/module.go [hunk 26]
```diff
@@ -122,20 +123,29 @@ func CreateModule(options *Options) (*Module, error) {
...
+				Line:     o.Pos.Line,
 			},
-		})
+		}
+		if options.OutputValues {
+			terraformOutputs, err := loadOutputValues(options)
+			if err != nil {
+				log.Fatal(err)
+			}
+			output.Value = terraformOutputs[output.Name].Value
```
[the remark this answers is not in the document]
Comment (gshel) [reply internal/pkg/tfconf/module.go]: That's so true! Great catch

File: internal/pkg/tfconf/output.go (hunk 9 above)
Comment (khos2ow) [reply internal/pkg/tfconf/output.go]: So the behavior seems to be as:

- terraform 0.12.x: `type` of output is always `[]interface{}`
- terraform 0.11.x and prior: `type` of outputs is always `string`

In our case it doesn't make any difference at the moment because we're not showing the _type_ of output and we only need to extract it properly, which by using `Type interface{}` in `TerraformOutput` struct this will be fine.

File: internal/pkg/tfconf/output.go [hunk 27]
```diff
@@ -34,3 +35,10 @@ func (a outputsSortedByPosition) Swap(i, j int) {
 func (a outputsSortedByPosition) Less(i, j int) bool {
 	return a[i].Position.Filename < a[j].Position.Filename || a[i].Position.Line < a[j].Position.Line
 }
+
+// TerraformOutput is used for unmarshalling `terraform outputs --json` into
+type TerraformOutput struct {
+	Sensitive bool        `json:"sensitive"`
+	Type      string      `json:"type"`
```
[the remark this answers is not in the document]
Comment (gshel) [reply internal/pkg/tfconf/output.go]: Perhaps at some point in the future, something cool can be done with this!

File: internal/pkg/testutil/testing.go [hunk 28]
```diff
@@ -9,23 +9,30 @@ import (
...
 	path, err := getExampleFolder()
 	if err != nil {
-		return nil, "", err
-	}
-	options := &tfconf.Options{
-		Path: path,
+		return nil, err
 	}
+	options.Path = path
+	options.OutputValuesPath = path + options.OutputValuesPath
```
[the remark this answers is not in the document]
Comment (gshel) [reply internal/pkg/testutil/testing.go]: Shoot I didn't remove the trailing slash from my tests. Everything seemed to pass though. Should I update the tests to remove them?

File: internal/pkg/tfconf/module.go (hunk 25 above)
Comment (khos2ow) [reply internal/pkg/tfconf/module.go]: I've had some problem with this to make it work, but here is the working snippet

```go
func loadOutputValues(options *Options) (map[string]*TerraformOutput, error) {
	var out []byte
	var err error

	if options.OutputValuesPath == "" {
		cmd := exec.Command("terraform", "output", "-json")
		cmd.Dir = options.Path
		if out, err = cmd.Output(); err != nil {
			return nil, fmt.Errorf("caught error while reading the terraform outputs: %v", err)
		}
	} else {
		if out, err = ioutil.ReadFile(options.OutputValuesPath); err != nil {
			return nil, fmt.Errorf("caught error while reading the terraform outputs file at %s: %v", options.OutputValuesPath, err)
		}
	}

	var terraformOutputs map[string]*TerraformOutput
	err = json.Unmarshal(out, &terraformOutputs)
	if err != nil {
		return nil, err
	}
	return terraformOutputs, err
}
```

As you can see I'm not using `options.OutputValuesPath` at all unless it was explicitly passed by the user, and as such I suggest to remove the default value for it in `cmd/root.go` as well.

Comment (gshel) [reply internal/pkg/tfconf/module.go]: I'll make that change this morning!
